File: spepcpp/src/spep/sessions/impl/SessionCacheImpl.cpp
/* Copyright 2006-2007, Queensland University of Technology
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not 
 * use this file except in compliance with the License. You may obtain a copy of 
 * the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * Author: Shaun Mangelsdorf
 * Creation Date: 08/01/2007
 * 
 * Purpose: Implements the session cache object, of which only one exists per SPEP.
 */

#include "spep/sessions/impl/SessionCacheImpl.h"

#include "saml2/exceptions/InvalidParameterException.h"
#include "spep/exceptions/InvalidStateException.h"
#include "spep/exceptions/InvalidSessionIdentifierException.h"

#include <iterator>

#include "spep/UnicodeStringConversion.h"

#include <unicode/calendar.h>

#include <boost/date_time/posix_time/posix_time.hpp>


spep::SessionCacheImpl::SessionCacheImpl( spep::ReportingProcessor *reportingProcessor )
:
_localReportingProcessor( reportingProcessor->localReportingProcessor( "spep::SessionCacheImpl" ) ),
_sessionIDs(),
_esoeSessions(),
_unauthenticatedSessions()
{}

void spep::SessionCacheImpl::getPrincipalSession(spep::PrincipalSession &principalSession, std::string sessionID)
{
	ScopedLock lock(_principalSessionsMutex);
	
	// Find operation on map returns an iterator which either points at end() or 
	// at a kv pair which is the requested object.
	SessionIDMap::const_iterator iter( _sessionIDs.find( sessionID ) );
	if ( iter != _sessionIDs.end() )
	{
		// From the SessionIDMap we now have the esoe session identifier.
		// So we can look up the value and return it now.
		ESOESessionMap::const_iterator esoeIter( _esoeSessions.find( iter->second ) );
		if ( esoeIter == _esoeSessions.end() )
		{
			// Serious problem. Inconsistent map. No ESOE session ID for the given session.
			
			this->_localReportingProcessor.log( FATAL, "Session cache has become inconsistent. No ESOE session found for principal session, even though the session ID was found successfully." );
			throw InvalidStateException( "Session cache has become inconsistent" );
		}
		
		principalSession = esoeIter->second;
		
		// If the session hasn't expired, return it.
		if( principalSession.getSessionNotOnOrAfter() > boost::posix_time::second_clock::universal_time() )
		{
			return;
		}
		
		this->_localReportingProcessor.log( DEBUG, "Session expiry: " + sessionID + " was set to expire at " + boost::posix_time::to_simple_string(principalSession.getSessionNotOnOrAfter()) );
		
		// Expired. Terminate it.
		this->terminatePrincipalSession( principalSession.getESOESessionID() );
	}
	
	// Session wasn't found
	throw InvalidSessionIdentifierException();
}

void spep::SessionCacheImpl::getPrincipalSessionByEsoeSessionID(spep::PrincipalSession &principalSession, std::wstring esoeSessionID)
{
	ScopedLock lock(_principalSessionsMutex);
	
	ESOESessionMap::const_iterator iter( _esoeSessions.find( esoeSessionID ) );
	if ( iter != _esoeSessions.end() )
	{
		// Already have the ESOE session identifier. No need to retrieve it.
		std::pair<std::wstring, PrincipalSession> principalSessionPair = *iter;
		principalSession = principalSessionPair.second;
		
		// If the session hasn't expired, return it.
		if( principalSession.getSessionNotOnOrAfter() > boost::posix_time::second_clock::universal_time() )
		{
			return;
		}
		
		// Expired. Terminate it.
		// It would have at least 1 session ID in the list.. so .. terminate using that.
		this->terminatePrincipalSession( esoeSessionID );
	}
	
	// Session wasn't found
	throw InvalidSessionIdentifierException();
}

void spep::SessionCacheImpl::insertPrincipalSession( std::string sessionID, spep::PrincipalSession &principalSession )
{
	ScopedLock lock(_principalSessionsMutex);
	
	if (principalSession.getESOESessionID().length() == 0)
	{
		SAML2LIB_INVPARAM_EX( "Principal session had no ESOE session identifier" );
	}
	
	ESOESessionMap::iterator esoeIter( _esoeSessions.find( principalSession.getESOESessionID() ) );
	// Check if this is a new session..
	if ( esoeIter == _esoeSessions.end() )
	{	
		// Insert the principal session into the esoe session map.
		std::pair<ESOESessionMap::iterator, bool> result = 
			_esoeSessions.insert( std::make_pair( principalSession.getESOESessionID(), principalSession ) );
			
		// Ensure that the insert operation worked.
		if (! result.second)
		{
			SAML2LIB_INVPARAM_EX( "Failed to insert the principal session" );
		}
		
		// Make sure the session object contains the local session ID so the session gets cleaned up properly.
		bool found = false;
		std::vector<std::string> &principalSessionIDList = result.first->second.getSessionIDList();
		for( std::vector<std::string>::iterator sessionIDIter = principalSessionIDList.begin();
			sessionIDIter != principalSessionIDList.end(); ++sessionIDIter )
		{
			if( sessionIDIter->compare( sessionID ) == 0 )
			{
				found = true;
				break;
			}
		}
		
		if( !found )
		{
			principalSessionIDList.push_back( sessionID );
		}
	}
	// otherwise it's a new index for a session we already know about.
	else
	{
		PrincipalSession::ESOESessionIndexMapType::iterator esoeSessionIndexMapIterator;
		for( esoeSessionIndexMapIterator = principalSession.getESOESessionIndexMap().begin();
			esoeSessionIndexMapIterator != principalSession.getESOESessionIndexMap().end();
			++esoeSessionIndexMapIterator )
		{
			std::wstring esoeSessionIndex( esoeSessionIndexMapIterator->first );
			std::string localSessionID( esoeSessionIndexMapIterator->second );
			esoeIter->second.addESOESessionIndexAndLocalSessionID( esoeSessionIndex, localSessionID );
		}
	}
	
	// Insert into the map with SessionID as key
	std::pair<SessionIDMap::iterator, bool> secondResult =
		_sessionIDs.insert( std::make_pair( sessionID, principalSession.getESOESessionID() ) );
}

void spep::SessionCacheImpl::terminatePrincipalSession(const std::wstring esoeSessionID)
{
	ScopedLock lock(_principalSessionsMutex);
	
	// Loop up the principal session by ESOE session ID
	ESOESessionMap::iterator esoeIter = _esoeSessions.find(esoeSessionID);
	if ( esoeIter != _esoeSessions.end() )
	{
		// Go through the list of all local session identifiers for that principal session
		// and terminate them all.
		for( PrincipalSession::SessionIDListType::iterator sessionIDIter = esoeIter->second.getSessionIDList().begin();
			sessionIDIter != esoeIter->second.getSessionIDList().end();
			++sessionIDIter )
		{
			std::string sessionID( *sessionIDIter );
			SessionIDMap::iterator removeIter = _sessionIDs.find( sessionID );
			
			if( removeIter != _sessionIDs.end() )
			{
				_sessionIDs.erase( removeIter );
			}
		}
		
		// After we're done, remove the ESOE session ID from the session cache.
		_esoeSessions.erase( esoeIter );
	}
}

void spep::SessionCacheImpl::getUnauthenticatedSession(spep::UnauthenticatedSession &unauthenticatedSession, std::wstring requestID)
{
	ScopedLock lock(_unauthenticatedSessionsMutex);
	
	// Get the unauthenticated session by SAML request ID.
	UnauthenticatedSessionMap::const_iterator iter = _unauthenticatedSessions.find( requestID );
	if (iter != _unauthenticatedSessions.end())
	{
		unauthenticatedSession = iter->second;
		
		return;
	}
	
	throw InvalidSessionIdentifierException();
}

void spep::SessionCacheImpl::insertUnauthenticatedSession(spep::UnauthenticatedSession &unauthenticatedSession)
{
	ScopedLock lock(_unauthenticatedSessionsMutex);
	
	if (unauthenticatedSession.getAuthnRequestSAMLID().length() == 0)
	{
		SAML2LIB_INVPARAM_EX( "Unauthenticated session had no AuthnRequest SAML ID associated with it." );
	}
	
	std::pair< UnauthenticatedSessionMap::iterator, bool > result =
		_unauthenticatedSessions.insert( std::make_pair( unauthenticatedSession.getAuthnRequestSAMLID(), unauthenticatedSession ) );
	// If the insert failed the second element of the result pair will be false.
	if (!result.second)
	{
		SAML2LIB_INVPARAM_EX( "Failed to insert the unauthenticated session" );
	}
}

void spep::SessionCacheImpl::terminateUnauthenticatedSession(std::wstring requestID)
{
	ScopedLock lock(_unauthenticatedSessionsMutex);
	
	// Find by request ID and terminate
	UnauthenticatedSessionMap::iterator iter = _unauthenticatedSessions.find( requestID );
	if (iter != _unauthenticatedSessions.end())
	{
		_unauthenticatedSessions.erase(iter);
	}
}

void spep::SessionCacheImpl::terminateExpiredSessions( int sessionCacheTimeout )
{
	{
		ScopedLock lock(_principalSessionsMutex);
		
		for( ESOESessionMap::iterator iter = _esoeSessions.begin(); iter != _esoeSessions.end(); /* increment in loop body */ )
		{
			// If the session hasn't expired, skip it.
			if( iter->second.getSessionNotOnOrAfter() > boost::posix_time::second_clock::universal_time() )
			{
				++iter;
				continue;
			}
			// Otherwise, remove it.
			else
			{
				// This works because the iterator is incremented before the 
				// erase() call is made, but the old value is still passed in.
				_esoeSessions.erase( iter++ );
			}
		}
	}
	{
		ScopedLock lock(_unauthenticatedSessionsMutex);
		
		for( UnauthenticatedSessionMap::iterator iter = _unauthenticatedSessions.begin(); iter != _unauthenticatedSessions.end(); /* increment in loop body */ )
		{
			// If the session hasn't expired, skip it.
			if( iter->second.getIdleTime() < sessionCacheTimeout )
			{
				++iter;
				continue;
			}
			// Otherwise, remove it.
			else
			{
				// This works because the iterator is incremented before the 
				// erase() call is made, but the old value is still passed in.
				_unauthenticatedSessions.erase( iter++ );
			}
		}
	}
}

File: spepjava/src/com/qut/middleware/spep/sessions/impl/SessionCacheImpl.java
/* 
 * Copyright 2006, Queensland University of Technology
 * Licensed under the Apache License, Version 2.0 (the "License"); you may not 
 * use this file except in compliance with the License. You may obtain a copy of 
 * the License at 
 * 
 *   http://www.apache.org/licenses/LICENSE-2.0 
 * 
 * Unless required by applicable law or agreed to in writing, software 
 * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT 
 * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the 
 * License for the specific language governing permissions and limitations under 
 * the License.
 * 
 * Author: Shaun Mangelsdorf / Bradley Beddoes
 * Creation Date: 13/11/2006 / 03/03/2007
 * 
 * Purpose: Implements the SessionCache interface.
 */
package com.qut.middleware.spep.sessions.impl;

import java.text.MessageFormat;
import java.util.Date;
import java.util.GregorianCalendar;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Vector;
import java.util.Map.Entry;
import java.util.concurrent.locks.ReentrantLock;

import javax.xml.datatype.XMLGregorianCalendar;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import com.qut.middleware.spep.sessions.Messages;
import com.qut.middleware.spep.sessions.PrincipalSession;
import com.qut.middleware.spep.sessions.SessionCache;
import com.qut.middleware.spep.sessions.UnauthenticatedSession;
import com.qut.middleware.spep.util.CalendarUtils;

/** */
public class SessionCacheImpl implements SessionCache
{
	protected Map<String, PrincipalSession> sessions;
	protected Map<String, PrincipalSession> esoeSessions;
	protected Map<String, UnauthenticatedSession> unauthenticatedSessions;
	private CleanupThread cleanupThread;
	protected long sessionCacheTimeout;
	protected long sessionCacheInterval;

	private ReentrantLock lock;

	protected void lock()
	{
		this.lock.lock();
	}

	protected void unlock()
	{
		this.lock.unlock();
	}

	/* Local logging instance */
	private Logger logger = LoggerFactory.getLogger(SessionCacheImpl.class.getName());

	/**
	 * Default constructor.
	 * 
	 * @param reportingProcessor
	 *            The processor to use for logging.
	 * @param sessionCacheTimeout
	 *            Expiry age in seconds for cached sessions. Sessions older than specified time will be removed. from
	 *            the cache at regular intervals.
	 * @param sessionCacheInterval
	 *            interval in seconds session cache between polls to remove expired entries.
	 */
	public SessionCacheImpl(long sessionCacheTimeout, long sessionCacheInterval)
	{
		if (sessionCacheTimeout > Long.MAX_VALUE / 1000)
		{
			throw new IllegalArgumentException(Messages.getString("SessionCacheImpl.13")); //$NON-NLS-1$
		}
		if (sessionCacheInterval > Long.MAX_VALUE / 1000)
		{
			throw new IllegalArgumentException(Messages.getString("SessionCacheImpl.14")); //$NON-NLS-1$
		}

		this.sessionCacheTimeout = sessionCacheTimeout * 1000;
		this.sessionCacheInterval = sessionCacheInterval * 1000;

		this.sessions = new HashMap<String, PrincipalSession>();
		this.esoeSessions = new HashMap<String, PrincipalSession>();
		this.unauthenticatedSessions = new HashMap<String, UnauthenticatedSession>();

		this.cleanupThread = new CleanupThread();
		this.cleanupThread.start();

		this.lock = new ReentrantLock(true);

		this.logger.info(Messages.getString("SessionCacheImpl.0")); //$NON-NLS-1$
	}
	
	public void cleanup()
	{
		this.cleanupThread.stopRunning();
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.qut.middleware.spep.sessions.SessionCache#getClientSession(java.lang.String)
	 */
	public PrincipalSession getPrincipalSession(String sessionID)
	{
		PrincipalSession principalSession;

		lock();
		try
		{
			principalSession = this.sessions.get(sessionID);
		}
		finally
		{
			unlock();
		}

		if (principalSession != null)
		{
			XMLGregorianCalendar thisXmlCal = CalendarUtils.generateXMLCalendar();
			GregorianCalendar thisCal = thisXmlCal.toGregorianCalendar();

			if (thisCal.getTime().before(principalSession.getSessionNotOnOrAfter()))
			{
				this.logger.debug("Continuing with cached session " + principalSession.getEsoeSessionID() + " current time is: " + thisCal.getTime() + " sessionNotOnAfter was set to: " + principalSession.getSessionNotOnOrAfter());
				return principalSession;
			}

			this.logger.info("Terminating session " + principalSession.getEsoeSessionID() + " current time is: " + thisCal.getTime() + " sessionNotOnAfter was set to: " + principalSession.getSessionNotOnOrAfter());
			this.logger.debug(MessageFormat.format(Messages.getString("SessionCacheImpl.9"), principalSession.getEsoeSessionID())); //$NON-NLS-1$

			terminatePrincipalSession(principalSession);
		}

		this.logger.debug(MessageFormat.format(Messages.getString("SessionCacheImpl.8"), sessionID)); //$NON-NLS-1$

		return null;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.qut.middleware.spep.sessions.SessionCache#getClientSessionByEsoeSessionIndex(java.lang.String)
	 */
	public PrincipalSession getPrincipalSessionByEsoeSessionID(String esoeSessionID)
	{
		PrincipalSession principalSession;

		lock();
		try
		{
			principalSession = this.esoeSessions.get(esoeSessionID);
		}
		finally
		{
			unlock();
		}

		if (principalSession != null)
		{
			XMLGregorianCalendar thisXmlCal = CalendarUtils.generateXMLCalendar();
			GregorianCalendar thisCal = thisXmlCal.toGregorianCalendar();

			if (thisCal.getTime().before(principalSession.getSessionNotOnOrAfter()))
			{
				this.logger.info("Continuing with cached session " + principalSession.getEsoeSessionID() + " current time is: " + thisCal.getTime() + " sessionNotOnAfter was set to: " + principalSession.getSessionNotOnOrAfter());
				return principalSession;
			}

			this.logger.info("Terminating session " + principalSession.getEsoeSessionID() + " current time is: " + thisCal.getTime() + " sessionNotOnAfter was set to: " + principalSession.getSessionNotOnOrAfter());
			this.logger.debug(MessageFormat.format(Messages.getString("SessionCacheImpl.9"), principalSession.getEsoeSessionID())); //$NON-NLS-1$

			terminatePrincipalSession(principalSession);
		}

		return null;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.qut.middleware.spep.sessions.SessionCache#putClientSession(java.lang.String,
	 *      com.qut.middleware.spep.sessions.PrincipalSession)
	 */
	public void putPrincipalSession(String sessionID, PrincipalSession principalSession)
	{
		if (principalSession.getEsoeSessionID() == null)
		{
			this.logger.error(MessageFormat.format(Messages.getString("SessionCacheImpl.1"), sessionID)); //$NON-NLS-1$
			return;
		}

		lock();
		try
		{
			this.sessions.put(sessionID, principalSession);
			this.esoeSessions.put(principalSession.getEsoeSessionID(), principalSession);
		}
		finally
		{
			unlock();
		}

		this.logger.debug(MessageFormat.format(Messages.getString("SessionCacheImpl.2"), sessionID)); //$NON-NLS-1$
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.qut.middleware.spep.sessions.SessionCache#terminateClientSession(java.lang.String)
	 */
	public void terminatePrincipalSession(PrincipalSession principalSession)
	{
		lock();
		try
		{
			/* Terminate all SPEP sessionID's that reference this principal */
			for (String sessionID : principalSession.getSessionIDList())
			{
				this.sessions.remove(sessionID);
			}

			this.esoeSessions.remove(principalSession.getEsoeSessionID());
			this.logger.debug(MessageFormat.format(Messages.getString("SessionCacheImpl.3"), principalSession.getEsoeSessionID())); //$NON-NLS-1$		
		}
		finally
		{
			unlock();
		}
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.qut.middleware.spep.sessions.SessionCache#terminateIndividualPrincipalSession(com.qut.middleware.spep.sessions.PrincipalSession,
	 *      java.lang.String)
	 */
	public void terminateIndividualPrincipalSession(PrincipalSession principalSession, String esoeSessionIndex)
	{
		lock();
		try
		{
			Map<String, String> sessionIndex = principalSession.getEsoeSessionIndex();
			if (sessionIndex != null)
			{
				String localSessionID = sessionIndex.get(esoeSessionIndex);

				if (localSessionID != null)
				{
					/* Remove this session from local sessionID cache */
					this.sessions.remove(localSessionID);
					principalSession.getEsoeSessionIndex().remove(esoeSessionIndex);

					/* If the principal has no further local mappings then terminate their ESOE mapping */
					if (principalSession.getEsoeSessionIndex().size() == 0)
					{
						this.esoeSessions.remove(principalSession.getEsoeSessionID());
					}

					this.logger.debug(MessageFormat.format(Messages.getString("SessionCacheImpl.3"), principalSession.getEsoeSessionID())); //$NON-NLS-1$
				}
			}
		}
		finally
		{
			unlock();
		}
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.qut.middleware.spep.sessions.SessionCache#getUnauthenticatedSession(java.lang.String)
	 */
	public UnauthenticatedSession getUnauthenticatedSession(String requestID)
	{
		UnauthenticatedSession unauthenticatedSession;
		lock();
		try
		{
			unauthenticatedSession = this.unauthenticatedSessions.get(requestID);
		}
		finally
		{
			unlock();
		}

		if (unauthenticatedSession != null)
			unauthenticatedSession.updateTime();

		return unauthenticatedSession;
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.qut.middleware.spep.sessions.SessionCache#putUnauthenticatedSession(java.lang.String,
	 *      com.qut.middleware.spep.sessions.UnauthenticatedSession)
	 */
	public void putUnauthenticatedSession(String requestID, UnauthenticatedSession unauthenticatedSession)
	{
		lock();
		try
		{
			this.unauthenticatedSessions.put(requestID, unauthenticatedSession);
		}
		finally
		{
			unlock();
		}

		unauthenticatedSession.updateTime();

		this.logger.debug(MessageFormat.format(Messages.getString("SessionCacheImpl.4"), requestID)); //$NON-NLS-1$
	}

	/*
	 * (non-Javadoc)
	 * 
	 * @see com.qut.middleware.spep.sessions.SessionCache#terminateUnauthenticatedSession(java.lang.String)
	 */
	public void terminateUnauthenticatedSession(String requestID)
	{
		lock();
		try
		{
			this.unauthenticatedSessions.remove(requestID);
		}
		finally
		{
			unlock();
		}

		this.logger.debug(MessageFormat.format(Messages.getString("SessionCacheImpl.5"), requestID)); //$NON-NLS-1$
	}

	/**
	 * Cleans up unauthenticated sessions which have timed out dependant on some local setting. Authenticated sessions
	 * are terminated based on the time SessionNotOnOrAfter value set by the ESOE
	 * 
	 */
	private class CleanupThread extends Thread
	{
		/* Local logging instance */
		private Logger logger = LoggerFactory.getLogger(CleanupThread.class.getName());
		private boolean running;

		protected CleanupThread()
		{
			super("SPEP Session Cache cleanup thread"); //$NON-NLS-1$
		}

		@Override
		public void run()
		{
			this.running = true;
			while (this.running)
			{
				try
				{
					Thread.sleep(SessionCacheImpl.this.sessionCacheInterval);

					this.logger.debug(Messages.getString("SessionCacheImpl.10")); //$NON-NLS-1$
					cleanup();
				}
				catch (Exception e)
				{
					this.logger.error(MessageFormat.format(Messages.getString("SessionCacheImpl.11"), e.getMessage())); //$NON-NLS-1$
				}
			}
		}
		
		public void stopRunning()
		{
			this.running = false;
			this.interrupt();
		}

		private void cleanup()
		{
			List<String> expired = new Vector<String>();

			lock();
			try
			{
				this.logger.info(Messages.getString("SessionCacheImpl.15")); //$NON-NLS-1$
				/* Remove principal sessions that have expired */
				for (String esoeID : SessionCacheImpl.this.esoeSessions.keySet())
				{
					PrincipalSession principal = SessionCacheImpl.this.esoeSessions.get(esoeID);
					if (principal != null)
					{
						XMLGregorianCalendar thisXmlCal = CalendarUtils.generateXMLCalendar();
						GregorianCalendar thisCal = thisXmlCal.toGregorianCalendar();

						if (thisCal.getTime().after(principal.getSessionNotOnOrAfter()))
						{
							this.logger.info("Terminating session " +principal.getEsoeSessionID() + " current time is: " + thisCal + " sessionNotOnAfter was set to: " + principal.getSessionNotOnOrAfter());
							this.logger.debug(Messages.getString("SessionCacheImpl.16") + principal.getEsoeSessionID() + Messages.getString("SessionCacheImpl.17")); //$NON-NLS-1$ //$NON-NLS-2$
							terminatePrincipalSession(principal);
						}
					}
				}

				// Now clean up the unauthenticated sessions
				for (Entry<String, UnauthenticatedSession> entry : SessionCacheImpl.this.unauthenticatedSessions.entrySet())
				{
					if (entry.getValue().getIdleTime() > SessionCacheImpl.this.sessionCacheTimeout)
					{
						expired.add(entry.getKey());
					}
				}

				for (String requestID : expired)
				{
					this.logger.debug(Messages.getString("SessionCacheImpl.18") + requestID + Messages.getString("SessionCacheImpl.19")); //$NON-NLS-1$ //$NON-NLS-2$
					terminateUnauthenticatedSession(requestID);
				}
			}
			finally
			{
				unlock();
			}
		}
	}
}
